### PR TITLE
Bluetooth: GATT: Initial GATT BSIM tests

### PR DESCRIPTION
**Pull request by Thalley**
Adds initial GATT BSIM tests for the client
and server functionality, testing simple
reads and writes.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]

**Comment by martintv** [comment]
I see some tests in tests/bluetooth/bsim_bt use printk and some use LOG_, what's the preferred way really? LOG_ seems better to me and I see you use printk in this PR...   but I don't fully understand the difference, other than that LOG_ adds the module_name:function_name before each message, which is nice

File: tests/bluetooth/bsim_bt/bsim_test_gatt/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+if (NOT DEFINED ENV{BSIM_COMPONENTS_PATH})
+	message(FATAL_ERROR "This test requires the BabbleSim simulator. Please set\
+ the  environment variable BSIM_COMPONENTS_PATH to point to its components \
+ folder. More information can be found in\
+ https://babblesim.github.io/folder_structure_and_env.html")
+endif()
+
+find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})
+project(bsim_test_advx)
```
Comment (martintv) [inline tests/bluetooth/bsim_bt/bsim_test_gatt/CMakeLists.txt:13]: copy-paste error?

File: tests/bluetooth/bsim_bt/bsim_test_gatt/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources} )
+
+zephyr_include_directories(
+  $ENV{BSIM_COMPONENTS_PATH}/libUtilv1/src/
+  $ENV{BSIM_COMPONENTS_PATH}/libPhyComv1/src/
+  ${ZEPHYR_BASE}/subsys/bluetooth/controller/include
```
Comment (martintv) [inline tests/bluetooth/bsim_bt/bsim_test_gatt/CMakeLists.txt:21]: do you realy need controller/include for a gatt test?

**Comment by cvinayak** [comment]
> I see some tests in tests/bluetooth/bsim_bt use printk and some use LOG_, what's the preferred way really? LOG_ seems better to me and I see you use printk in this PR... but I don't fully understand the difference, other than that LOG_ adds the module_name:function_name before each message, which is nice

Applications/samples use `printk` as user interaction, compared to `LOG_` which are debug messages used in subsystems code that can be disabled (compile time and/or runtime).

**Comment by martintv** [comment]
.

**Review by martintv — APPROVED** [review]

**Review by CarbuneanuAlexandru — COMMENTED** [review]
what about a very short description for this test ?

**Comment by Thalley** [comment]
> what about a very short description for this test ?

Can you elaborate a bit on what you'd like?

**Comment by CarbuneanuAlexandru** [comment]
> 
I guess I haven seen in many tests here, I am speaking about a short explanation about the purpose of the test and the pass/fail criteria

**Comment by Thalley** [comment]
> > 
> 
> I guess I haven seen in many tests here, I am speaking about a short explanation about the purpose of the test and the pass/fail criteria

Hmm, in that case I believe in-code documentation might be best instead of a README, as the file(s) may contain several test cases with different pass criteria etc. 

There's already a little bit here: https://github.com/zephyrproject-rtos/zephyr/pull/41630/files#diff-6a9223c69bdf3097c14e8466f155cf6703430ad007e44754cb919c5dd859000eR5-R7 but isn't very specific.

**Review by CarbuneanuAlexandru — APPROVED** [review]

**Comment by CarbuneanuAlexandru** [comment]
he you consider to put more stuff un common.c file, functions like connected(), disconnected() and others... and leave in gatt_client_test.c and gatt_server_test.c only the test cases and little more? That would help a lot against code duplication and will increase reusability

File: tests/bluetooth/bsim_bt/bsim_test_gatt/src/gatt_server_test.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		return BT_GATT_ERR(BT_ATT_ERR_SUCCESS);
+	}
+
+	(void)memcpy(long_chrc_data + offset, buf, len);
+	prepare_count = 0;
+
+	return len;
+}
+
+/* Eddystone Configuration Service Declaration */
```
Comment (asbjornsabo) [inline tests/bluetooth/bsim_bt/bsim_test_gatt/src/gatt_server_test.c:127]: Eddystone?

**Review by asbjornsabo — APPROVED** [review]

**Comment by Thalley** [comment]
> Did you consider to put more stuff un common.c file, functions like connected(), disconnected() and others... and leave in gatt_client_test.c and gatt_server_test.c only the test cases and little more? That would help a lot against code duplication and will increase reusability

For the LE Audio tests we actually do have a little more in the common files. 

Moving stuff connected and disconnected into a common file, would also require a bit of rework in how the flags are used, but is doable. The disconnected callback is different from central to peripheral (only central should do `bt_conn_unref` for some reason), but it is possible by adding a simple `if role == central` check. 

That being said, the common code is about 50 lines, so it's not a huge gain to put those in common, when you think about that you'll also need to move the flags etc. to make that work.

**Pull request merged**